myia/anf_ir.py

```diff
@@ -274,6 +307,20 @@ def __init__(self, graph: Graph) -> None:
         """Construct the parameter."""
         super().__init__([], PARAMETER, graph)
 
+    def __str__(self) -> str:
+        """Return a string representation of this node.
+
+        The debugging `name` attribute will be used to output strings of the
+        form `x` if possible. Otherwise it will default to `arg0` where `0` is
+        the argument index.
+
+        """
+        if hasattr(self.debug, 'name'):
+            return self.debug.name
+        if self in self.graph.parameters:
+            return f"arg{self.graph.parameters.index(self)}"
```

Comment:
Under this scheme, an optimization that drops an argument would transform `f(arg0, arg1, arg2)` into `f(arg0, arg1)` regardless of which argument is dropped. That means optimizations can accidentally "rename" parameters and that could be confusing in some circumstances. I would prefer parameter names to be set in stone, although that does make avoiding collisions a little harder.

Reply:
Not sure what I think about this. Some things come to mind:

* Naming should be optional, since it is only for debugging. I don't want to have to give each variable a unique name when I'm just messing about in `pdb`, or be forced to deal with parameter naming when I'm quickly trying out an optimization that affects parameters. You can hide these requirements behind helper functions of course, but I'd like to avoid needing helper functions to instantiate nodes.
* Unique naming requires some form of scoping and/or globals, since each node name needs to be aware of all the other nodes that it cannot clash with and their names. Ew.

Given that, I am inclined to say:

* `__str__` should handle a lack of names gracefully
* But we can have our own parser and optimizations fix the name of each parameter, so that the parameter renaming problem you describe doesn't happen unless people write custom parsers/optimizations.

Reply:
New nodes can be named automatically either with the `uuid` package, or with a global monotonically increasing integer, or more simply by setting their name as `${id(parameter)}`. Also, insofar that the names are not significant in code generation, it's also fine (and more readable) to set them to random dictionary words and just ignore the small collision probability.